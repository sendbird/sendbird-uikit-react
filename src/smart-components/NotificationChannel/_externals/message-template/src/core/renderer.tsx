import React from 'react';

import type { BasicProps, ComponentType, ComponentsUnion, GetProperties } from '../types/components';

type FC<Props, ParsedStyle> = (props: BasicProps<Props, ParsedStyle>) => React.ReactElement | null;

type ComponentProperties<Type extends ComponentType> = GetProperties<Type, ComponentsUnion>;

interface RendererOptions<ParsedStyle> {
  views?: {
    [component in ComponentType]?: FC<ComponentProperties<component>, ParsedStyle>;
  };
}

export type Renderer<ParsedStyle> = {
  [component in ComponentType]: FC<ComponentProperties<component>, ParsedStyle>;
};

const FRAGMENT = ({ children }: React.PropsWithChildren) => <React.Fragment>{children}</React.Fragment>;

export function createRenderer<ParsedStyle = object | string>(
  opts?: RendererOptions<ParsedStyle>,
): Renderer<ParsedStyle> {
  return {
    box: opts?.views?.box || FRAGMENT,
    text: opts?.views?.text || FRAGMENT,
    image: opts?.views?.image || FRAGMENT,
    imageButton: opts?.views?.imageButton || FRAGMENT,
    textButton: opts?.views?.textButton || FRAGMENT,
  };
}
