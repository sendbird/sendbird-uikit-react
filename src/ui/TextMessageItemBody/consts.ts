export const TEXT_MESSAGE_BODY_CLASSNAME = 'sendbird-text-message-item-body';
