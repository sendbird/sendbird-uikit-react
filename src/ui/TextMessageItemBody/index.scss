@import '../../styles/variables';

.sendbird-text-message-item-body {
  position: relative;
  display: inline-block;
  box-sizing: content-box;
  padding: 8px 12px;
  border-radius: 16px;
  &.reactions {
    border-radius: 16px 16px 0px 0px;
  }
  &.incoming {
    @include themed() { background-color: t(bg-1) }
  }
  &.outgoing {
    @include themed() { background-color: t(primary-3) }
  }

  &.mouse-hover, &:hover {
    &.incoming {
      @include themed() { background-color: t(bg-2) }
    }
    &.outgoing {
      @include themed() { background-color: t(primary-4) }
    }
  }

  .sendbird-text-message-item-body__message {
    white-space: pre-wrap;
    word-break: break-all;
  }
}
