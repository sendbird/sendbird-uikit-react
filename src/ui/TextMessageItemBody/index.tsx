import './index.scss';
import React, { ReactElement, useContext, useMemo } from 'react';
import type { UserMessage } from '@sendbird/chat/message';

import Label, { LabelTypography, LabelColors } from '../Label';
import { getClassName, isEditedMessage } from '../../utils';
import { LocalizationContext } from '../../lib/LocalizationContext';
import uuidv4 from '../../utils/uuid';
import Word from '../Word';

interface Props {
  className?: string | Array<string>;
  message: UserMessage;
  isByMe?: boolean;
  mouseHover?: boolean;
  isMentionEnabled?: boolean;
  isReactionEnabled?: boolean;
}

export default function TextMessageItemBody({
  className,
  message,
  isByMe = false,
  mouseHover = false,
  isMentionEnabled = false,
  isReactionEnabled = false,
}: Props): ReactElement {
  const { stringSet } = useContext(LocalizationContext);
  const isMessageMentioned = isMentionEnabled && message?.mentionedMessageTemplate?.length > 0 && message?.mentionedUsers?.length > 0;
  const sentences: Array<Array<string>> = useMemo(() => {
    if (isMessageMentioned) {
      return message?.mentionedMessageTemplate?.split(/\n/).map((sentence) => sentence.split(/\s/));
    }
    return message?.message?.split(/\n/).map((sentence) => sentence.split(/\s/));
  }, [message?.mentionedMessageTemplate]);
  return (
    <Label
      type={LabelTypography.BODY_1}
      color={isByMe ? LabelColors.ONCONTENT_1 : LabelColors.ONBACKGROUND_1}
    >
      <div className={getClassName([
        className,
        'sendbird-text-message-item-body',
        isByMe ? 'outgoing' : 'incoming',
        mouseHover ? 'mouse-hover' : '',
        (isReactionEnabled && message?.reactions?.length > 0) ? 'reactions' : '',
      ])}>
        {
          sentences.map((sentence, index) => {
            return [
              sentence.map((word) => {
                return (
                  <Word
                    key={uuidv4()}
                    word={word}
                    message={message}
                    isByMe={isByMe}
                  />
                );
              }),
              sentences?.[index + 1] ? <br /> : null,
            ]
          })
        }
        {
          isEditedMessage(message) && (
            <Label
              className="sendbird-text-message-item-body__message edited"
              type={LabelTypography.BODY_1}
              color={isByMe ? LabelColors.ONCONTENT_2 : LabelColors.ONBACKGROUND_2}
            >
              {` ${stringSet.MESSAGE_EDITED} `}
            </Label>
          )
        }
      </div>
    </Label>
  );
}
