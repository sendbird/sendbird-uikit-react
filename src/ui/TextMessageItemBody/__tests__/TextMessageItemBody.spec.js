import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import TextMessageItemBody from "../index";

const createMockMessage = (process) => {
  const mockMessage = {
    message: '',
    updatedAt: 0,
    reactions: [],
    messageType: 'user',
    isUserMessage: () => true,
  };
  return process ? process(mockMessage) : mockMessage;
};

describe('TextMessageItemBody', () => {
  it('should have class names of own basic status', () => {
    const insertingClassName = 'test-class-name';
    const component = mount(
      <TextMessageItemBody
        className={insertingClassName}
        message={createMockMessage()}
      />
    );
    expect(
      component.find(`.${insertingClassName}`).hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-text-message-item-body').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-text-message-item-body.outgoing').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-text-message-item-body.incoming').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-text-message-item-body.mouse-hover').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-text-message-item-body.reactions').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-text-message-item-body__message').hostNodes().exists()
    ).toBe(false);
    expect(
      component.find('.sendbird-text-message-item-body__message.edited').hostNodes().exists()
    ).toBe(false);
  });

  it('should have class name by isByMe prop', () => {
    const outgoingMessage = mount(
      <TextMessageItemBody
        message={createMockMessage()}
        isByMe
      />
    );
    expect(
      outgoingMessage.find('.sendbird-text-message-item-body.outgoing').hostNodes().exists()
    ).toBe(true);
    expect(
      outgoingMessage.find('.sendbird-text-message-item-body.incoming').hostNodes().exists()
    ).toBe(false);
    const incomingMessage = mount(
      <TextMessageItemBody
        message={createMockMessage()}
        isByMe={false}
      />
    );
    expect(
      incomingMessage.find('.sendbird-text-message-item-body.outgoing').hostNodes().exists()
    ).toBe(false);
    expect(
      incomingMessage.find('.sendbird-text-message-item-body.incoming').hostNodes().exists()
    ).toBe(true);
  });

  it('should have class name by mouse hover prop', () => {
    const component = mount(
      <TextMessageItemBody
        message={createMockMessage()}
        mouseHover
      />
    );
    expect(
      component.find('.sendbird-text-message-item-body.mouse-hover').hostNodes().exists()
    ).toBe(true);
    expect(
      component.find('.sendbird-text-message-item-body.incoming.mouse-hover').hostNodes().exists()
    ).toBe(true);
  });

  it('should have class name by reactions of message prop', () => {
    const component = mount(
      <TextMessageItemBody
        message={createMockMessage((mockMessage) => ({
          ...mockMessage,
          reactions: [{}, {}, {}],
        }))}
      />
    );
    expect(
      component.find('.sendbird-text-message-item-body.reactions').hostNodes().exists()
    ).toBe(true);
  });

  it('should have words component by split message', () => {
    const messageText = 'First second third fourth fifth';
    const component = mount(
      <TextMessageItemBody
        message={createMockMessage((mockMessage) => ({
          ...mockMessage,
          message: messageText,
        }))}
      />
    );
    expect(
      component.find('.sendbird-text-message-item-body__message.edited').hostNodes().exists()
    ).toBe(false);
    const editedMessage = mount(
      <TextMessageItemBody
        message={createMockMessage((mockMessage) => ({
          ...mockMessage,
          message: messageText,
          updatedAt: 10101010,
        }))}
      />
    );
    expect(
      editedMessage.find('.sendbird-text-message-item-body__message.edited').hostNodes().exists()
    ).toBe(true);
  });

  it('should do a snapshot test of the TextMessageItemBody DOM', function () {
    const component = renderer.create(
      <TextMessageItemBody
        className="class-name-for-snapshot"
        message={createMockMessage((mock) => ({
          ...mock,
          message: 'First second third fourth fifth',
          updatedAt: 1010,
          reactions: [{}, {}, {}],
        }))}
        isByMe
        mouseHover
      />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
