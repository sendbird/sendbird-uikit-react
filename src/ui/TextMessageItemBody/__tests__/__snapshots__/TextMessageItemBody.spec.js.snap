// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/TextMessageItemBody should do a snapshot test of the TextMessageItemBody DOM 1`] = `
<DocumentFragment>
  <span
    class="sendbird-label sendbird-label--body-1 sendbird-label--color-oncontent-1"
  >
    <div
      class="class-name-for-snapshot sendbird-text-message-item-body outgoing mouse-hover "
    >
      <span
        class="sendbird-word"
      >
        First
      </span>
      <span
        class="sendbird-word"
      >
        second
      </span>
      <span
        class="sendbird-word"
      >
        third
      </span>
      <span
        class="sendbird-word"
      >
        fourth
      </span>
      <span
        class="sendbird-word"
      >
        fifth
      </span>
      <span
        class="sendbird-text-message-item-body__message edited sendbird-label sendbird-label--body-1 sendbird-label--color-oncontent-2"
      >
         (edited) 
      </span>
    </div>
  </span>
</DocumentFragment>
`;
